numpy/core/shape_base.py

```diff
@@ -423,32 +434,112 @@ def format_index(index):
         return parent_index, _nx.ndim(arrays)
 
 
-def _block(arrays, max_depth, result_ndim):
+def _concatenate_shapes(shapes, axis):
+    """Given array shapes, return the resulting shape that would occur
+    after array concatenation.
+
+    concatenate(arrs, axis).shape == _concatenate_shapes([a.shape for a in arrs], axis)
     """
-    Internal implementation of block. `arrays` is the argument passed to
-    block. `max_depth` is the depth of nested lists within `arrays` and
-    `result_ndim` is the greatest of the dimensions of the arrays in
-    `arrays` and the depth of the lists in `arrays` (see block docstring
-    for details).
+    # Take a shape, any shape
+    first_shape = shapes[0]
+    for shape in shapes[1:]:
+        if (shape[:axis] != first_shape[:axis] or
+                shape[axis+1:] != first_shape[axis+1:]):
+            raise ValueError('Mismatched array shapes in block.')
+    shape_on_dim = sum(shape[axis] for shape in shapes)
+    return first_shape[:axis] + (shape_on_dim,) + first_shape[axis+1:]
+
+
+def _atleast_nd(a, ndim):
+    # Ensures `a` has at least `ndim` dimensions by prepending
+    # ones to `a.shape` as necessary
+    return array(a, ndmin=ndim, copy=False, subok=True)
+
+
+def _accumulate(values):
+    # Helper function because Python 2.7 doesn't have
+    # itertools.accumulate
+    value = 0
+    accumulated = []
+    for v in values:
+        value += v
+        accumulated.append(value)
+
+    return tuple(accumulated)
+
+
+def _concatenate_shapes_as_slices(shapes, axis):
     """
-    def atleast_nd(a, ndim):
-        # Ensures `a` has at least `ndim` dimensions by prepending
-        # ones to `a.shape` as necessary
-        return array(a, ndmin=ndim, copy=False, subok=True)
-
-    @recursive
-    def block_recursion(self, arrays, depth=0):
-        if depth < max_depth:
-            if len(arrays) == 0:
-                raise ValueError('Lists cannot be empty')
-            arrs = [self(arr, depth+1) for arr in arrays]
-            return _nx.concatenate(arrs, axis=-(max_depth-depth))
-        else:
-            # We've 'bottomed out' - arrays is either a scalar or an array
-            # type(arrays) is not list
-            return atleast_nd(arrays, result_ndim)
+    Produce the slices that match the destination each array would end up
+    along axis.
+
+    For the following situation::
 
-    return block_recursion(arrays)
+        ret = concatenate([a, b, c], axis=0)
+        sl_a, sl_b, sl_c = concatenate_slices([a, b, c], axis=0)
+
+    The result will be that ``ret[sl_a] == a``, ``ret[sl_b] == b``, and
+    ``ret[sl_c] == c``.
+    Similar properties hold for other values of ``axis``.
+
+    See Also
+    --------
+        concatenate : Join a sequence of arrays together.
```

Comment:
Doubt these lines are worth it - this is an internal function, and I'd hope the connection to `concatenate` is obvious